text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,776 @@
+# Programmatic Toolkit
+
+* **Original Author(s):**: @mrgrain
+* **Tracking Issue**: #300
+* **API Bar Raiser**: @rix0rrr
+
+The [Programmatic Toolkit] allows integrators to build lifecycle management solutions for CDK applications.
+It provides them with full control over all output and allows to interweave actions with custom steps.
+
+## Working Backwards
+
+### README
+
+[Programmatic Toolkit] is a library to run CDK actions programmatically.
+The package provides implementations for the following actions:
+
+- `bootstrap`
+- `list`
+- `synth`
+- `deploy`
+- ...
+
+#### Self-contained example
+
+```ts
+// Create a [Programmatic Toolkit]
+await using cdk = new Toolkit({
+  ioHost: new CloudWatchLogsIoHost('my-log-stream'), // optional, e.g. an IoHost to write any output to CloudWatch Logs
+});
+
+// Define the CloudExecutable
+const cx = CloudExecutable.fromCloudAssemblyProducer(async (context) => {
+  const app = new cdk.App({ context });
+  const stack = new cdk.Stack(app);
+  return app.synth();
+});
+
+// Synth and store a cached CloudExecutable, this won't run synth again
+const cxCache = await cdk.synth(cx);
+
+// Deploy our pre-synth'ed app
+const deployment = await cdk.deploy(cxCache, { stacks: ['MyTestStack'] });
+
+// Returns a list of ARNs of the deployment
+return deployment.listResources().filter(r => r.arn);
```

Comment:
Is there the possibility of filtering resources here by node path? This might make it easier to locate specific resources for a system under test.

Reply:
This is a great suggestion! Might not be in the first version, but should be easy to add.